source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -302,7 +373,57 @@ const absl::string_view DnsFilter::getClusterNameForDomain(const absl::string_vi
   return {};
 }
 
-bool DnsFilter::resolveViaClusters(DnsQueryContextPtr& context, const DnsQueryRecord& query) {
+bool DnsFilter::resolveClusterService(DnsQueryContextPtr& context, const DnsQueryRecord& query) {
+  size_t cluster_endpoints = 0;
+
+  // Get the service_list config for the domain
+  const auto* service_config = getServiceConfigForDomain(query.name_);
+  if (service_config != nullptr) {
+
+    // We can redirect to more than one cluster, but only one is supported
+    const auto& cluster_target = service_config->targets_.begin();
+    const auto& target_name = cluster_target->first;
+    const auto& attributes = cluster_target->second;
+
+    // Determine if there is a cluster
+    Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(target_name);
+    if (cluster == nullptr) {
+      ENVOY_LOG(trace, "No cluster found for service target: {}", target_name);
+      return false;
+    }
+
+    // Add a service record for each cluster endpoint using the cluster name
+    const std::chrono::seconds ttl = getDomainTTL(target_name);
+    for (const auto& hostsets : cluster->prioritySet().hostSetsPerPriority()) {
+      for (const auto& host : hostsets->hosts()) {
+
+        // Create the service record element and increment the SRV record answer count
+        auto config = std::make_unique<DnsSrvRecord>(service_config->name_, service_config->proto_,
+                                                     service_config->port_, service_config->ttl_);
+
+        config->addTarget(target_name, attributes.priority, attributes.weight);
```

Comment:
This is where I would expect the port number from the cluster to be used

Reply:
Yep.  That makes sense.